### DISCUSSION TITLE
Getting cursor position from TSLexer?

### DISCUSSION THREAD

**Discussion by osa1** [tree-sitter/tree-sitter#1038] [Q&A] [answered]
In languages usually identifiers and keywords have the same lexical syntax. So it makes sense to collect characters in a lexer, and check if the collected id is a keyword, return keyword if it is, return identifier otherwise.

Currently with tree-sitter external lexer API the only way to implement this is by collecting the seen characters into a buffer, then comparing it with "if", "else", etc. (keywords in the language).

This is fine but it allocates redundantly. If tree-sitter lexer API gave me access to the cursor index I could do `&source[cursor_start..cursor_end] == <keyword>`, without allocating any intermediate strings.

Is something like this possible today?

**Comment by ahlinc** [comment, marked as answer]
Tree-sitter doesn't provide access to its internal read buffer cause this would also require to expose all machinery about the buffer maintenance, an API to check its boundary and to read the next input chunk. All this would eliminate simplicity and the simple comparison as you'd like still can't be possible for an any buffer state. The internal lexer's scanner logic uses the same API like exposed to external scanners.
Also it's needed to think about possibility of relatively long tokens, making their buffering before comparison isn't effective way how to process them. Current API allows to compare by a symbol and to fail fast if it's obvious that further reading and comparing have no sence.
Another note is that for static tokens usually there is no need to use the external scanner to distinguish them from identifiers, tree-sitter has a builtin ability for this, see the `word` parameter in the tree-sitter's [DSL](https://tree-sitter.github.io/tree-sitter/creating-parsers#the-grammar-dsl) documentation and a [Keyword Extraction](https://tree-sitter.github.io/tree-sitter/creating-parsers#keyword-extraction) feature description.

**Reply by osa1** [reply]
I don't understand how exposing the cursor position in `TSLexer`, perhaps with a function pointer similar to `get_column`, requires drastic changes.

> Tree-sitter doesn't provide access to its internal read buffer cause this would also require to expose all machinery about the buffer maintenance

I don't need its internal buffer, I already have access to the source. I don't understand why would it require to expose anything other than just the cursor position, in a way similar to `get_column`.

I didn't know about the `word` parameter. It's a bit strange to do keyword matching in parser, but it's workable.

**Reply by ahlinc** [reply]
A simplified example of the regular call chain:
`client code -> tree-sitter parser -> external scanner/lexer`

You have access for your source code only in the _client code_ and there you may have a position that is an offset from the begining. On the _external scanner_ side there is no possibility to have access to the source code if you write your grammar in a portable way that it can be used with any avalable tree-sitter binding for specific programming language. That is a reason that I see why tree-sitter doesn`t expose the offset in the source code to the external scanner.
